File: crates/rp1-freertos/tests/newlib_memcpy.c
/* Instruction-level Cortex-M3 model check; this is not RP1 hardware evidence. */
#include <stddef.h>
#include <stdint.h>

extern void *memcpy(void *, const void *, size_t);
extern void __aeabi_memcpy(void *, const void *, size_t);
extern void __aeabi_memcpy4(void *, const void *, size_t);
extern void __aeabi_memcpy8(void *, const void *, size_t);
static volatile uint8_t source[576] __attribute__((aligned(8)));
static volatile uint8_t destination[576] __attribute__((aligned(8)));
static volatile unsigned current_kind, current_length, current_src, current_dst;

static void semihost(unsigned operation, const void *argument) {
    register unsigned r0 __asm__("r0") = operation;
    register const void *r1 __asm__("r1") = argument;
    __asm__ volatile("bkpt 0xab" : "+r"(r0), "+r"(r1) : : "memory", "cc");
}

static void hex(unsigned value) {
    char text[10];
    for (unsigned i = 0; i < 8; ++i)
        text[i] = "0123456789abcdef"[(value >> (28 - 4 * i)) & 15];
    text[8] = ' '; text[9] = 0;
    semihost(4, text);
}

__attribute__((noreturn)) static void finish(unsigned status) {
    const unsigned arguments[] = {0x20026, status};
    semihost(0x20, arguments);
    for (;;) {}
}

static void failure_context(void) {
    semihost(4, "kind,length,src,dst(hex)=");
    hex(current_kind); hex(current_length); hex(current_src); hex(current_dst);
    semihost(4, "\n");
}

void fault(void) {
    semihost(4, "TRAP cfsr(hex)=");
    hex(*(volatile unsigned *)0xe000ed28);
    semihost(4, "\n");
    failure_context();
    finish(2);
}

void reset(void) {
    *(volatile unsigned *)0xe000ed14 = (1u << 9) | (TRAP_MODE << 3);
    __asm__ volatile("dsb\nisb" : : : "memory");
    unsigned cases = 0;
    for (unsigned kind = 0; kind < 4; ++kind) {
        unsigned alignment = kind < 2 ? 1 : (kind == 2 ? 4 : 8);
        for (unsigned index = 0; index < 260; ++index) {
            unsigned length = index < 258 ? index : index + 253;
            for (unsigned so = 0; so < 4; ++so) {
                for (unsigned d = 0; d < 4; ++d) {
                    unsigned start_src = 16 + so * alignment;
                    unsigned start_dst = 16 + d * alignment;
                    current_kind = kind; current_length = length;
                    current_src = start_src; current_dst = start_dst;
                    for (unsigned i = 0; i < sizeof(source); ++i) {
                        source[i] = (i * 37 + 11) & 127;
                        destination[i] = 0xa5;
                    }
                    void *to = (void *)&destination[start_dst];
                    const void *from = (const void *)&source[start_src];
                    unsigned bad = 0;
                    switch (kind) {
                    case 0: bad = memcpy(to, from, length) != to; break;
                    case 1: __aeabi_memcpy(to, from, length); break;
                    case 2: __aeabi_memcpy4(to, from, length); break;
                    case 3: __aeabi_memcpy8(to, from, length); break;
                    }
                    for (unsigned i = 0; i < sizeof(source); ++i) {
                        unsigned wanted = i >= start_dst && i < start_dst + length
                            ? ((start_src + i - start_dst) * 37 + 11) & 127 : 0xa5;
                        bad |= destination[i] != wanted;
                        bad |= source[i] != ((i * 37 + 11) & 127);
                    }
                    if (bad) {
                        semihost(4, "FAIL copy/return/canary/source\n");
                        failure_context();
                        finish(1);
                    }
                    ++cases;
                }
            }
        }
    }
    semihost(4, "PASS cases(hex)="); hex(cases); semihost(4, "\n");
    finish(0);
}

__attribute__((section(".vectors"), used))
const uintptr_t vectors[] = {
    0x20010000, (uintptr_t)reset, (uintptr_t)fault, (uintptr_t)fault,
    (uintptr_t)fault, (uintptr_t)fault, (uintptr_t)fault,
};

File: crates/rp1-freertos/tests/newlib_memcpy.py
#!/usr/bin/env python3
"""Test only build.rs-pinned newlib memcpy members on QEMU, never on RP1."""
import hashlib
import os
from pathlib import Path
import re
import shlex
import subprocess
import tempfile

HERE = Path(__file__).resolve().parent
CC = os.environ.get("RP1_FREERTOS_CC", "arm-none-eabi-gcc")
AR = os.environ.get("RP1_FREERTOS_AR", "arm-none-eabi-ar")
FLAGS = ["-mcpu=cortex-m3", "-mthumb", "-mfloat-abi=soft"]


def run(*args, **kwargs):
    return subprocess.run(args, check=True, capture_output=True, timeout=15, **kwargs).stdout


build = (HERE.parent / "build.rs").read_text()
version = re.search(r'const GCC_VERSION: &str = "([^"]+)"', build)[1]
assert run(CC, "-dumpmachine").strip() == b"arm-none-eabi"
assert run(CC, "-dumpfullversion").decode().strip() == version == "14.2.1"
archive = run(CC, *FLAGS, "-print-file-name=libc.a").decode().strip()
members = re.findall(r'\(\s*"(memcpy|aeabi_memcpy)",\s*"([0-9a-f]{64})",', build)
assert {name for name, _ in members} == {"memcpy", "aeabi_memcpy"}
assert b"mps2-an385" in run("qemu-system-arm", "-machine", "help")
print(run(CC, "--version").decode().splitlines()[0])
print(run("qemu-system-arm", "--version").decode().splitlines()[0])
print(f"archive={archive}")
with tempfile.TemporaryDirectory(prefix="newlib-memcpy-") as temporary:
    work = Path(temporary)
    objects = []
    for name, expected in members:
        member = f"libc_a-{name}.o"
        data = run(AR, "p", archive, member)
        digest = hashlib.sha256(data).hexdigest()
        assert digest == expected, f"changed newlib member: {member}"
        obj = work / member
        obj.write_bytes(data)
        objects.append(str(obj))
        undefined = run("arm-none-eabi-nm", "-u", str(obj)).decode().strip()
        assert undefined == ("U memcpy" if name == "aeabi_memcpy" else "")
        print(f"{member} sha256={digest} undefined={undefined or 'none'}")
    linker = work / "test.ld"
    linker.write_text("""ENTRY(reset)
MEMORY { FLASH (rx) : ORIGIN = 0, LENGTH = 4M
         RAM (rwx) : ORIGIN = 0x20000000, LENGTH = 64K }
SECTIONS { .text : { KEEP(*(.vectors)) *(.text*) *(.rodata*) } > FLASH
           .data : { *(.data*) } > RAM
           .bss (NOLOAD) : { *(.bss*) *(COMMON) } > RAM }
""")
    flags = FLAGS + ["-std=c11", "-O2", "-ffreestanding", "-fno-builtin",
                     "-fno-tree-loop-distribute-patterns", "-fno-unwind-tables",
                     "-fno-asynchronous-unwind-tables", "-Wall", "-Wextra", "-Werror"]
    print("compile flags=" + shlex.join(flags))
    for trap in (0, 1):
        test_object, elf = work / "test.o", work / "test.elf"
        run(CC, *flags, f"-DTRAP_MODE={trap}", "-c", str(HERE / "newlib_memcpy.c"),
            "-o", str(test_object))
        undefined = run("arm-none-eabi-nm", "-u", str(test_object)).decode().split()
        assert undefined == ["U", "__aeabi_memcpy", "U", "__aeabi_memcpy4",
                             "U", "__aeabi_memcpy8", "U", "memcpy"]
        run(CC, *FLAGS, "-nostdlib", "-Wl,--build-id=none", f"-Wl,-T,{linker}",
            str(test_object), *objects, "-o", str(elf))
        assert not run("arm-none-eabi-nm", "-u", str(elf)).strip()
        command = ["qemu-system-arm", "-M", "mps2-an385", "-cpu", "cortex-m3", "-nic", "none",
                   "-nographic", "-monitor", "none", "-serial", "none",
                   "-semihosting-config", "enable=on,target=native", "-kernel", str(elf)]
        result = subprocess.run(command, capture_output=True, text=True, timeout=15)
        output = result.stdout + result.stderr
        print(f"CCR.UNALIGN_TRP={trap} exit={result.returncode}\n{output}", end="")
        if trap == 0:
            assert result.returncode == 0 and "PASS cases(hex)=00004100" in output
        else:
            # Pinned assembly requires unaligned accesses; this is a diagnostic, not a pass.
            assert result.returncode == 2 and "TRAP cfsr(hex)=01000000" in output
            assert "kind,length,src,dst(hex)=00000000 00000002 00000010 00000011" in output
    print("PASS main matrix: 16640 cases; trap-on diagnostic trapped as reported.")
print("Temporary objects/ELFs removed. Instruction-level model only; no RP1 hardware evidence.")
